src/runtime_src/core/pcie/driver/linux/include/mailbox_proto.h

```diff
@@ -346,14 +346,15 @@ struct xcl_mailbox_clock_freqscaling {
 #define XCL_MB_REQ_FLAG_REQUEST		(1 << 1)
 /**
  * struct mailbox_req - mailbox request message header
- * @req: opcode
+ * @req: opcode (see enum xcl_mailbox_request)
  * @flags: flags of this message
  * @data: payload of variable length
  */
 struct xcl_mailbox_req {
 	uint64_t flags;
-	enum xcl_mailbox_request req;
-	char data[1]; /* variable length of payload */
+	uint32_t req;
+	uint32_t padding;
+	char data[0]; /* variable length of payload */
```

Comment:
This change will also require change in xclmgmt2 driver on host in Azure. This will require MS to modify the driver and update on all the hosts in all the Data Centers across the world. Although this is not an immediate impact, as this change doesn't affect 2022.1, but going forward it would impact them without adding additional functionality for the current Azure deployment. I am not oppose to the change, if it makes thing simpler for X3 firmware, but someone have to get the buy-in from MS, as to the impact going forward. 
Also, any reason why char data[1] was changed to data[0]? With current change, it won't compile with Visual Studio compilers. Please move it back to data[1].

Reply:
> In cloud use cases, the version of xclmgmt running in host may not be the same as xocl which runs in guest (such as in Azure).

I submitted the patch in the hope that the situation described wouldn't be a valid usecase for the drivers.  If it is, then I agree that the patch in its current form cannot be applied.  Removing the new padding member completely would help, but Arpit raises issues with the data[0] change.

> This change will also require change in xclmgmt2 driver on host in Azure... someone have to get the buy-in from MS, as to the impact going forward.

This sounds impractical - is that the point you're making here?  If so, then it's another large, probably impassible barrier.

> Also, any reason why char data[1] was changed to data[0]? With current change, it won't compile with Visual Studio compilers. Please move it back to data[1].

The data[0] is part of the fix for the issue.  As described in the CR, the driver allocates its buffers using calculations of the form `sizeof(struct xcl_mailbox_req) + message_length_bytes`, (with the structure size being 16) but starts storing the message inside the data member of `struct xcl_mailbox_req` instead of at the end (at byte offset 12), which leaves 4 bytes at the end of the buffer unused.  These four bytes are awkward to handle in the X3 FW because we don't have the luxury of buffering full messages in RAM like the host drivers - we have to process message data in packets at a time, and expect the last packet of the message to be the end of the data.   With this padding at the end of the message, that expectation does not hold.  Changing to data[0] gobbles up those 4 bytes and makes FW's life easier.

However, if needs to compile for Windows too, we need to find another way.  I've been thinking about this and another way to gobble the extra 4 bytes and maintain compatibility between older and newer XRT drivers would be to change how the drivers calculate their buffer sizes i.e. instead of modifying the structure definition and breaking things, we could leave it alone and change `sizeof(struct xcl_mailbox_req) + message_length_bytes` into `offsetof(struct xcl_mailbox_req, data) + message_length_bytes`  This would eliminate the padding at the end (making FW's life easy) and keep the message data position in the buffer consistent between different drivers.  The only difference is that the buffers allocated by the drivers would be 4 bytes smaller, but since drivers receiving messages 1) learn the necessary buffer size from the first packet and 2) access the data by casting it to `struct xcl_mailbox_req*` and dereferencing the data member, things should work ok.

The main disadvantage of this approach is that every request buffer sizing statement in the drivers would need to be fixed, so the patch to fix the issue would touch more places.  It would also look a little unconventional.

The main question from me now is whether this approach is acceptable to you and Max?

Reply:
In the meantime, I've given proposal a shot and it works with X3, so I've updated this PR with a replacement patch.  Please let me know if you think this revised solution is acceptable.